Fix race condition after attaching tenant with branches.

hlinnaka opened the pull request:
After tenant attach, it's possible that the branch point of a branch is beyond the parent's last-record-LSN. Need to wait for it to arrive, like GetPage does on the child branch.

@NanoBjorn encountered this while working on tenant migration. After migrating a tenant with a parent and child branch, connecting to the child branch failed with an error like:

```
FATAL:  "base/16385" is not a valid data directory
DETAIL:  File "base/16385/PG_VERSION" is missing.
```

This commit adds a test case that reproduced the bug, although with a slightly different error message. I'm pretty sure you could reproduce that exact error by fiddling with the LSN where the branch is created, but this covers the underlying bug just the same.

hlinnaka commented:
Hmm, this might not be a full fix: is it possible that the ancestor timeline is still in Attaching state, when a child already starts to receive WAL from the safekeepers or to serve GetPage requests? In that case, 'wait_lsn' returns an error. I think we change all the timelines to Active state while holding a lock on the tenant's state, so that probably cannot happen. But I'm not 100% sure. If it happens, the walreceiver or getpage request will get an error, but will succeed when it's retried later. So I think this is fine for now, but another case to consider when we tighten up the tenant/timeline lifecycle management.

koivunej commented on test_runner/regress/test_tenant_detach.py:
```diff
@@ -108,6 +108,59 @@ def test_tenant_reattach(
+        current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
+    env.neon_cli.create_branch("child", tenant_id=tenant_id, ancestor_start_lsn=current_lsn)
+
+    with main_endpoint.connect(dbname="neondb").cursor() as cur:
```
Python technical question: does this mean both of the `db connection` (from `connect(dbname=...)` and `cursor` are context objects which will be cleared on exit from `with` or will this leak the connections?

hlinnaka replied:
I'm not sure, but I realized that the `cursor` function can also take connection parameters as arguments. So I removed the connect() calls, making the question moot.

(In general I'm not to worried about leaking things during tests. They get cleaned up at end of test anyway)

koivunej replied:
Yeah for connections it is not important, but if you had chained the `endpoint.start()` (or something which provides that) and another method on that, it'd be interesting to know if this should be an antipattern none of our tools alert on or not. Leaking a compute would result in unwanted behaviour most likely.

koivunej reviewed (commented):
This seems to make sense to me, luckily @NanoBjorn found the case. I was thinking we should start to do random re-migrations to uncover issues like these.

koivunej commented on pageserver/src/tenant.rs:
```diff
@@ -3294,15 +3294,25 @@ mod tests {
-
+        // The branchpoints should contain all timelines, even ones marked
+        // as Broken.
         let branchpoints = &tline.gc_info.read().unwrap().retain_lsns;
         assert_eq!(branchpoints.len(), 1);
         assert_eq!(branchpoints[0], Lsn(0x40));
```
```suggestion
        {
            let branchpoints = &tline.gc_info.read().unwrap().retain_lsns;
            assert_eq!(branchpoints.len(), 1);
            assert_eq!(branchpoints[0], Lsn(0x40));
        }
```

NanoBjorn reviewed (commented):
jfyi this fix worked for me locally, so waiting for it main

hlinnaka commented on test_runner/regress/test_tenant_detach.py:
```diff
@@ -108,6 +108,59 @@ def test_tenant_reattach(
+    # create new nenant
+    tenant_id, timeline_id = env.neon_cli.create_tenant()
+
+    main_endpoint = env.endpoints.create_start("main", tenant_id=tenant_id)
+    with main_endpoint.cursor() as cur:
+        cur.execute("CREATE DATABASE neondb")
+        current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
+    env.neon_cli.create_branch("child", tenant_id=tenant_id, ancestor_start_lsn=current_lsn)
+
+    with main_endpoint.cursor(dbname="neondb") as cur:
+        cur.execute("CREATE TABLE test_migrate_one AS SELECT generate_series(1,100)")
+    main_endpoint.stop()
+
+    child_endpoint = env.endpoints.create_start("child", tenant_id=tenant_id)
+    with child_endpoint.cursor(dbname="neondb") as cur:
+        cur.execute("CREATE TABLE test_migrate_one AS SELECT generate_series(1,200)")
+    child_endpoint.stop()
+
+    pageserver_http.tenant_detach(tenant_id)
+    pageserver_http.tenant_attach(tenant_id)
```
Added a new test that is more explicit. And the new test actually hit the scenario that I envisioned at https://github.com/neondatabase/neon/pull/4170#issuecomment-1536882955; the logical size calculation tries to read from the parent timeline, before the parent timeline has become active. So I added a fix for that too.

problame replied:
Yeah, that new test is mostly good. But, _this_ test hasn't been failpoint-ized yet. Can we use the same failpoint / add another?

hlinnaka replied:
Ok, added the same failpoint mechanism to this test too.

problame commented on pageserver/src/walingest.rs:
```diff
@@ -305,6 +305,17 @@ impl<'a> WalIngest<'a> {
                 }
             }
+        } else if decoded.xl_rmid == pg_constants::RM_LOGICALMSG_ID {
+            let info = decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK;
+            if info == pg_constants::XLOG_LOGICAL_MESSAGE {
+                // This is a convenient way to make the WAL ingestion pause at
+                // particular point in the WAL. For more fine-grained control,
+                // we could peek into the message and only pause if it contains
+                // a particular string, for example, but this is enough for now.
+                info!("processing logical message");
+                fail::fail_point!("wal-ingest-logical-message");
+                info!("processing logical message: done");
+            }
         }
```
In the test, you are using the `sleep(...)` failpoint action.
That uses `std::thread::sleep()` internally.
But this is an async function.
So, it blocks the executor thread.

Please use `utils::failpoint_sleep_millis_async` and `return(milliseconds_to_sleep)` as a failpoint action in the test.


```suggestion
            let info = decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK;
            if info == pg_constants::XLOG_LOGICAL_MESSAGE {
                // This is a convenient way to make the WAL ingestion pause at
                // particular point in the WAL. For more fine-grained control,
                // we could peek into the message and only pause if it contains
                // a particular string, for example, but this is enough for now.
                info!("processing logical message");
                utils::failpoint_sleep_millis_async!(("wal-ingest-logical-message-sleep");
                info!("processing logical message: done");
            }
        }
```

problame commented on test_runner/regress/test_tenant_detach.py:
```diff
@@ -108,6 +108,130 @@ def test_tenant_reattach(
+# test reproduced one such case; the symptom was a 'FATAL: database
+# "neondb" does not exist' error on the child, when trying to connect
+# after re-attaching the tenant. The WAL record of CREATE DATABASE was
+# missed when searching the DbDirectory key in the ancestor.
+@pytest.mark.parametrize("remote_storage_kind", available_remote_storages())
+def test_tenant_with_branches_reattach_createdb(
+    neon_env_builder: NeonEnvBuilder,
+    remote_storage_kind: RemoteStorageKind,
+):
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_tenant_reattach",
+    )
```
```suggestion
# missed when searching the DbDirectory key in the ancestor.
@pytest.mark.parametrize("remote_storage_kind", available_remote_storages())
def test_tenant_with_branches_reattach_createdb(
    neon_env_builder: NeonEnvBuilder,
    remote_storage_kind: RemoteStorageKind,
):
    neon_env_builder.enable_remote_storage(
        remote_storage_kind=remote_storage_kind,
        test_name="test_tenant_with_branches_reattach_createdb",
    )
```

problame commented on test_runner/regress/test_tenant_detach.py:
```diff
@@ -108,6 +108,130 @@ def test_tenant_reattach(
+        cur.execute("SELECT * FROM test_reattach")
+        assert cur.fetchall() == [("before pause",), ("after pause",)]
+
+    # Sanity check: If this finished in under 5 s, the failpoint didn't sleep correctly.
+    assert time.time() - before_attach_time > 5
```
On an overloaded CI runner, this will be true even if we miss the failpoint.

Instead of checking timing, check the pageserver logs for presence / absence of the pg_logical_emit_message that you sent.

`assert not log_contains(the message)` before the attach to ensure we haven't yet processed the message at detach time. (Is this racy because the log is written asynchronously?)

Anyway, more importantly, at this time.time() line, use `wait_until(... log_contains(the message))` instead, to assert that we hit the failpoint.

You can keep the time-based check as a sanity-check, but, the above is more robust.

hlinnaka replied:
> Instead of checking timing, check the pageserver logs for presence / absence of the pg_logical_emit_message that you sent.

Ok, added.

> `assert not log_contains(the message)` before the attach to ensure we haven't yet processed the message at detach time.

Processing the WAL record doesn't print anything to the log, unless the failpoint is enabled.

> (Is this racy because the log is written asynchronously?)

Yes, it's a little racy. We have that problem with log_contains in general. Would be nice to fix, but not sure how, and hasn't caused any problems in practice so far.

> Anyway, more importantly, at this time.time() line, use `wait_until(... log_contains(the message))` instead, to assert that we hit the failpoint.

The SELECT should not return, until the sleep at the failpoint has finished. So we definitely should have already hit the failpoint at this point, there's no need to wait for it here. Otherwise something is broken.

problame commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -2218,6 +2258,36 @@ impl Timeline {
+                // yet nor re-streamed from the safekeepers. In other words,
+                // it's possible that the branch point is beyond the ancestor's
+                // last-record-LSN.
+                match ancestor.wait_to_become_active(ctx).await {
+                    Ok(()) => {}
+                    Err(state) if state == TimelineState::Stopping => {
+                        return Err(PageReconstructError::Stopping);
+                    }
+                    Err(state) => {
+                        return Err(PageReconstructError::Other(anyhow::anyhow!(
+                            "Timeline {} will not become active. Current state: {:?}",
+                            self.timeline_id,
+                            &state,
+                        )));
+                    }
+                }
```
1. I think the `Stopping` variant should be called `AncestorStopping`.
2. It should contain the ancestor timeline id that is stopping, and reproduce that timeline id in the error message.
3. The `Other(anyhow...())` message is using `self.timeline_id`, but, actually, it should use `ancestor.timeline_id`.

problame commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1845,12 +1873,24 @@ impl Timeline {
                     }
-                    x @ Err(_) => x.context("Failed to calculate logical size")?,
+                    Err(CalculateLogicalSizeError::Other(err)) => {
+                        if let Some(PageReconstructError::Stopping) =
+                            err.root_cause().downcast_ref()
+                        {
+                            // This can happen at tenant detach, when the parent timeline switches
+                            // to Stopping state while we're still calculating the initial timeline
+                            // size for the child. Like in the CalculateLogicalSizeError::Cancelled,
+                            // don't make noise.
+                            info!("initial size calculation failed because the timeline or its ancestor is Stopping, likely because the tenant is being detached");
+                            return Ok(());
+                        }
```
... or because we're restarting pageserver

Also, please print the error in the info! message using `{e:#}`.
If the comment above it is wrong about when this can happen, at least we'll have the error message.

Also, with my request to add a timeline_id field to the Stopping variant, printing the error will give us more breadcrumbs.

I think you can use the `if let Some(x @ PageReconstructError::Stopping(_))` syntax to get ahold of the error.

hlinnaka replied:
Currently, the PageReconstructError::Stopping doesn't carry extra information, so printing it  doesn't really add anything. I did it anyway, so that if we add extra information to it in the future, it becomes useful and we don't forget to add the printing at that point.

problame commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -926,6 +931,29 @@ impl Timeline {
         self.state.subscribe()
     }
 
+    pub async fn wait_to_become_active(
+        &self,
+        _ctx: &RequestContext, /* Prepare for use by cancellation */
+    ) -> Result<(), TimelineState> {
+        let mut receiver = self.state.subscribe();
+        loop {
+            let current_state = *receiver.borrow_and_update();
+            match current_state {
+                TimelineState::Loading => {
+                    // the sender cannot be dropped because we hold a refence to 'self'
+                    receiver.changed().await.expect("state sender dropped");
+                }
+                TimelineState::Active { .. } => {
+                    return Ok(());
+                }
+                TimelineState::Broken { .. } | TimelineState::Stopping => {
+                    // There's no chance the timeline can transition back into ::Active
+                    return Err(current_state);
+                }
+            }
+        }
+    }
```
Follow-up task (will create it and resolve this conversation then: investigate whether we can delete `connection_manager.rs wait_for_active_timeline` in favor of this one. This one is better and in the right module.)

problame replied:
https://github.com/neondatabase/neon/issues/4189

problame commented on libs/utils/src/lib.rs:
```diff
@@ -67,7 +67,7 @@ macro_rules! failpoint_sleep_millis_async {
             fail::fail_point!($name, |v: Option<_>| {
                 let millis = v.unwrap().parse::<u64>().unwrap();
-                Some(Duration::from_millis(millis))
+                Some(std::time::Duration::from_millis(millis))
```
NIT: I think leading `::` is the way to go here.
Same for all other types in this macro.

Or, `use ::std as std;` at the beginning of the block.

https://stackoverflow.com/a/65955962

@koivunej would probably know more.

hlinnaka replied:
This is being discussed at https://github.com/neondatabase/neon/pull/4195. I just needed something here to make the CI pass, but will commit PR #4195 before this.

problame commented on test_runner/regress/test_tenant_detach.py:
```diff
@@ -108,6 +108,147 @@ def test_tenant_reattach(
+
+    pageserver_http.tenant_detach(tenant_id)
+
+    # Wait before ingesting the WAL for CREATE DATABASE on the main branch. The original
+    # bug reproduced easily even without this, as there is always some delay between
+    # loading the timeline and establishing the connection to the safekeeper to stream and
+    # ingest the WAL, but let's make this less dependent on accidental timing.
+    pageserver_http.configure_failpoints([("wal-ingest-logical-message-sleep", "return(5000)")])
```
The `Wait before ingesting the WAL for CREATE DATABASE` is a lie, no?
It's just that, usually, the Python script is faster than the `CREATE DATABASE` WAL reaching pageservers.
Or am I missing something?

Also, for my understanding: before the fixes of this patch, what happened, exactly?
We create the child branch at the right LSN, but, didn't have the data.
So, the child branch asked safekeepers for the WAL of _what_ timeline? the child one or the parent one? And how did they respond?

hlinnaka replied:
Added a more detailed explanation. Let me know if it makes it clear or not

problame commented on test_runner/regress/test_tenant_detach.py:
```diff
@@ -108,6 +108,218 @@ def test_tenant_reattach(
+# None of these WAL records have been flushed to disk or uploaded to remote
+# storage in the pageserver yet, when the tenant is detached.
+#
+# After detach and re-attach, a walreceiver is spawned on both timelines.
+# They will connect to the safekeepers and start ingesting the WAL.
```
```suggestion
# After detach and re-attach, a walreceiver is spawned on both timelines.
# They will connect to the safekeepers and start ingesting the WAL
# from their respective IndexParts' `disk_consistent_lsn` onward.
```

problame reviewed (commented):
Thanks, your comment really helped me understand the bug at a deeper level.

Basically the summary is that we now, in a child branch, we back-fill the  wait for walreceiver to
back-fill the parent branch up to branch point LSN after re-attach.

But I now wonder whether it's the right approach to rely on SKs + walreceiver to back-fill the parent branch.

Isn't the real problem that we allowed branch creation for a branch LSN that is not yet remote consistent?
If we didn't allow that, then, instead of your fix to wait for ancestor's WAL to be re-streamed, we could
just assert that ancestor is at least at branch LSN.

The argument I'd make is that, detach+attach should be a pageserver-internal affair.
SKs availability should not impact ability to successfully detach+attach+access existing data.

hlinnaka commented:
> Thanks, your comment really helped me understand the bug at a deeper level.

Ok, great!

> Basically the summary is that we now, in a child branch, we back-fill the wait for walreceiver to back-fill the parent branch up to branch point LSN after re-attach.

I didn't understand that summary. What does back-filling mean? But I presume we're on the same page, anyway.

> But I now wonder whether it's the right approach to rely on SKs + walreceiver to back-fill the parent branch.

That's a fair question.

> Isn't the real problem that we allowed branch creation for a branch LSN that is not yet remote consistent? If we didn't allow that, then, instead of your fix to wait for ancestor's WAL to be re-streamed, we could just assert that ancestor is at least at branch LSN.
> 
> The argument I'd make is that, detach+attach should be a pageserver-internal affair. SKs availability should not impact ability to successfully detach+attach+access existing data.

I don't think there isn't anything wrong with the current approach per se either, but I agree it would feel less surprising.

It means that we need to wait for the remote storage uploads on the parent timeline to finish before the timeline creation can complete, but I guess that's acceptable.

We'd still need the "wait for ancestor timeline to become active" part of this PR anyway. Most of the code changes here were needed for that, the `wait_for_lsn` call is just a one-liner on top of that. So I'd suggest committing this PR as it is, and open a follow-up issue/PR to change timeline_create to wait for parent's remote uploads. What do you think?

koivunej commented on pageserver/src/http/routes.rs:
```diff
@@ -105,6 +105,9 @@ impl From<PageReconstructError> for ApiError {
                 ApiError::InternalServerError(anyhow::anyhow!("request was cancelled"))
             }
+            PageReconstructError::AncestorStopping(_) => {
+                ApiError::InternalServerError(anyhow::anyhow!(pre))
```
```suggestion
                ApiError::InternalServerError(anyhow::Error::new(pre))
```

Seems a bit weird otherwise to switch style between arms. Effect might be the same.

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -926,6 +935,29 @@ impl Timeline {
+            let current_state = *receiver.borrow_and_update();
+            match current_state {
+                TimelineState::Loading => {
+                    // the sender cannot be dropped because we hold a refence to 'self'
+                    receiver.changed().await.expect("state sender dropped");
```
I almost suggested "is this an expect on purpose" :)
```suggestion
                    receiver.changed().await.expect("holding a reference to self");
```

koivunej commented on test_runner/regress/test_tenant_detach.py:
```diff
@@ -108,6 +108,218 @@ def test_tenant_reattach(
+        current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
+
+    main_endpoint.stop()
+    env.neon_cli.create_branch("child", tenant_id=tenant_id, ancestor_start_lsn=current_lsn)
```
Before #4204 this should be racy with the detach; the `child` might not be uploaded to s3.

koivunej commented on test_runner/regress/test_tenant_detach.py:
```diff
@@ -108,6 +108,218 @@ def test_tenant_reattach(
+
+    # This fail point will pause the WAL ingestion on the main branch, after the
+    # the first insert
+    pageserver_http.configure_failpoints([("wal-ingest-logical-message-sleep", "return(5000)")])
```
What guarantees that the `main` hasn't already progressed past this? I think it's plausible that it'd be still in flight, but it might also not be.

koivunej replied:
This is probably a duplicate of https://github.com/neondatabase/neon/pull/4170/files#r1190986568

koivunej replied:
Because of the `detach` this works, so everything will be re-walreceived. Somehow I failed to "see the detach".

Perhaps here could be a `remote consistent lsn == 0` assertion? Though, the #4204 might make this much more difficult to hit.

hlinnaka replied:
Hmm, this does depend on the fact that the layer wasn't flushed and uploaded before the detach happens. If something caused the layers to be uploaded before the detach, or if we detach itself did it, this test would fail to reproduce the bug. The test includes a sanity check that the failpoint is hit after attach, so we would catch that.

koivunej reviewed (commented):
I think this still makes sense, and don't doubt the bug at all. I mentioned some known issues, and voiced a similar concern for not understanding what guarantees that we are late with WAL when the failpoint is activated. But I may have just understood why it works, so almost-approving.

koivunej commented on test_runner/regress/test_tenant_detach.py:
```diff
@@ -108,6 +108,218 @@ def test_tenant_reattach(
+#           <child branch is created>
+# 2. CREATE TABLE AS SELECT ...            3. CREATE TABLE AS SELECT ...
+#
+# None of these WAL records have been flushed to disk or uploaded to remote
+# storage in the pageserver yet, when the tenant is detached.
```
I think any freeze, flush, and wait uploads to complete, will be fine, because we are hoping that they are not yet on pageservers in-memory layers. I guess I'd prefer to have more asserts that `remote_consistent_lsn` is `initdb_lsn` (probably cannot be 0).

hlinnaka replied:
> IMO regular detach should freeze & flush, no?

/me reads the code

Huh, it does not. Detach calls `task_mgr::shutdown_tasks`, which requests the flush_loop to exit. Pageserver shutdown does freeze and flush, but not detach.

hlinnaka replied:
I think changing that is out of scope for this PR. But I'll change the test so that instead of detaching the tenant gracefully, it kills the pageserver and removes the tenant directory. That simulates attaching the tenant to a different pageserver, after a hard, unrecoverable crash of the original pageserver.

hlinnaka replied:
Opened https://github.com/neondatabase/neon/issues/4219 to track changing tenant_detach behavior

hlinnaka replied:
> I think any freeze, flush, and wait uploads to complete, will be fine, because we are hoping that they are not yet on pageservers in-memory layers. I guess I'd prefer to have more asserts that `remote_consistent_lsn` is `initdb_lsn` (probably cannot be 0).

To avoid the original bug, the WAL needs to be uploaded to remote storage. Otherwise, after re-attach the `remote_consistent_lsn` won't include it, and `get_page_reconstruct_data` won't see it, unless it waits for it to arrive.

Can you elaborate what assert you'd like to see, and where?

hlinnaka replied:
> As in if we had uploaded the layers to remote storage, then we wouldn't be walreceiving them anymore, we would just download the layers on-demand which is expected to work.

Correct!

hlinnaka replied:
Hmm, I'm not sure what value I would expect `remote_consistent_lsn` to have in the child here. The 'index_part.json` hasn't been uploaded at all yet, so even the existence of the timeline isn't durable in remote storage yet. But if you ignored that, then I guess the current behaviour is correct. Another valid answer IMO would be `remote_consistent_lsn==ancestor_start_lsn`.

https://github.com/neondatabase/neon/pull/4204 makes this more clear. With that, I think the correct answer here is `remote_consistent_lsn==ancestor_start_lsn`. Maybe add that assertion as part of that PR.

koivunej replied:
> `remote_consistent_lsn==ancestor_start_lsn`

Aah this is interesting, I didn't see it but obviously it should be like this. The asserts worked, but now I don't understand why. The timeline is branched off the lsn which should had been awaited for. 

I already merged the PR so I am doing a new one to clear up this confusion (of mine).

problame commented:
> I don't think there isn't anything wrong with the current approach per se either, but I agree it would feel less surprising.
> 
> It means that we need to wait for the remote storage uploads on the parent timeline to finish before the timeline creation can complete, but I guess that's acceptable.
> 
> We'd still need the "wait for ancestor timeline to become active" part of this PR anyway. Most of the code changes here were needed for that, the wait_for_lsn call is just a one-liner on top of that. So I'd suggest committing this PR as it is, and open a follow-up issue/PR to change timeline_create to wait for parent's remote uploads. What do you think?

Sure, we can take this PR now and fix branch creation later, although, as part of that later fix, IMO we should stop `wait_lsn` for ancestors and instead bail out if we don't have it, in the spirit of

>> The argument I'd make is that, detach+attach should be a pageserver-internal affair. SKs >> availability should not impact ability to successfully detach+attach+access existing data.

I'll create an issue for that.

---

On a side note, I'm now wondering whether we can currently branch in the future? Couldn't spot a check while skimming through `Tenant::branch_timeline` .

hlinnaka commented:
Updated:

- Changed the test so that it doesn't do detach anymore. Instead, it kills the pageserver and removes the tenant directory. This way, the test will continue to work even if we change graceful detach to upload all the data to S3 before completing. 
- moved the test to `test_tenant_relocation.py`, because it doesn't use detach anymore.

I split the commits so that one commit just moves the test without any other changes, and the other changes are in the next commit, so you can view the relevant difference more easily

problame commented on test_runner/regress/test_tenant_relocation.py:
```diff
@@ -512,3 +516,224 @@ def test_tenant_relocation(
+):
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_tenant_with_branches_reattach_slow_replay",
```
```suggestion
        test_name="test_emergency_relocate_with_branches_slow_replay",
```

problame commented on test_runner/regress/test_tenant_relocation.py:
```diff
@@ -512,3 +516,224 @@ def test_tenant_relocation(
+):
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_tenant_with_branches_reattach_createdb",
```
```suggestion
        test_name="test_emergency_relocate_with_branches_createdb",
```

problame reviewed (approved):
LGTM apart from few nitpicks & comment improvement
